### PR TITLE
Implemented ticket authentication.

### PR DESCRIPTION
**Pull request by fhriley**

### CODE REVIEW

**Comment by om26er** [comment]
Thanks for implementing ticket auth, I was actually planning to look into it as well, so I'd say great timing :)

This PR is actually changing(read: removing) a public API which needs input from @oberstet again. If its not too much of an effort, do you think you can separate out the ticket auth implementation only and overload the Client constructor to accept `List<IAuthenticator>` instead ?

We can then continue the discussion on changing `Client` constructor and `Client.add()`.

**Comment by fhriley** [comment]
I agree that changing a public API needs to be reviewed carefully. I would like to suggest that the public API change I made is the correct change. There are a couple reasons:

1. None of the authentication data is used in `Client`, it only used in `Session`. If we keep that data in `Client`, we need to add methods to `Session` to get it there from `Client`. Why not just put it in `Session` in the first place?

2. The API itself is suggesting the auth data belongs in `Session`:
```
public void add(Session session, String realm, List<IAuthenticator> authenticators)
```
In other words, you add that data with the `Session`, which suggests `Session` should be created with that data and hold it.

**Comment by fhriley** [comment]
Forking the API instead.
